2018 Robot/src/org/usfirst/frc/team4999/commands/autonomous/MoveDistance.java

```diff
@@ -0,0 +1,141 @@
+package org.usfirst.frc.team4999.commands.autonomous;
+
+import org.usfirst.frc.team4999.robot.MoPrefs;
+import org.usfirst.frc.team4999.robot.Robot;
+import org.usfirst.frc.team4999.robot.RobotMap;
+import org.usfirst.frc.team4999.robot.sensors.ADIS16448_IMU;
+import org.usfirst.frc.team4999.robot.sensors.VMXPi;
+import org.usfirst.frc.team4999.robot.subsystems.DriveSystem;
+
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.PIDController;
+import edu.wpi.first.wpilibj.PIDOutput;
+import edu.wpi.first.wpilibj.PIDSource;
+import edu.wpi.first.wpilibj.PIDSourceType;
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
+ */
+public class MoveDistance extends Command {
+	
+	private PIDController pid;
+	private Encoder left = RobotMap.leftDriveEncoder;
+	private Encoder right = RobotMap.rightDriveEncoder;
+	private double distance;
+	private static double ticksPerMeter = 1;
+	
+	
+	static class AverageEncoder implements PIDSource{
+		private Encoder left, right;
+
+		private PIDSourceType sourcetype = PIDSourceType.kDisplacement;
+
+
+		
+		public AverageEncoder(Encoder left, Encoder right) {
+			this.left = left;
+			this.right = right;
+		}
+		
+		@Override
+		public void setPIDSourceType(PIDSourceType pidSource) {
+			sourcetype = pidSource;
+		}
+
+		@Override
+		public PIDSourceType getPIDSourceType() {
+			return sourcetype;
+		}
+
+		@Override
+		public double pidGet() {
+			switch(sourcetype) {
+			case kRate:
+				return (left.getRate() + right.getRate()) / 2;
+			case kDisplacement:
+			default:
+				return (left.getDistance() + right.getDistance()) / 2;
+			}
+		}
+		
+	}	
+	
+
+	static class DriveAngleCorrect implements PIDOutput {
+		private DriveSystem output;
+		private Encoder left, right;
+		private double lStart, rStart;
+		VMXPi vmx = VMXPi.getInstance();
+		ADIS16448_IMU adis = new ADIS16448_IMU();
+		
+		private final double distanceBetweenWheels = 0.5; // meters
+		private final double moveErrGain = 0.1;
+		
+		public DriveAngleCorrect(DriveSystem output, Encoder left, Encoder right) {
+			this.output = output;
+			this.left = left;
+			this.right = right;
+			this.lStart = left.getDistance();
+			this.rStart = right.getDistance();
+		}
+		
+		 private double getAngle() { 
+			 if(vmx.getTimeSinceLastPacket() > vmx.RESPONSECUTOFF) {
+				return adis.getAngleZ();
+			 }
+			 return vmx.getAngle();
+	    }
+
+		@Override
+		public void pidWrite(double output) {
+			this.output.arcadeDrive(output, getAngle() * moveErrGain, RobotMap.auto_speed);
+		}
+		
+	}
+	
+	
+    public MoveDistance(double distance) {
+    	requires(Robot.driveSystem);
+    	this.distance = distance;
+    	MoPrefs prefs = MoPrefs.getInstance();
+    	pid = new PIDController(
```

Comment:
I'm not clear how a single PID controller can handle both linear movement and heading control at the same time, with the same PID constants.  What's going on here?